File: algorithm/merge-two-sorted-lists/merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode *rp = new ListNode(0),
                 *hp = rp;
        
        while (l1 && l2) {
            if (l1->val < l2->val) {
                hp->next = l1;
                l1 = l1->next;
            } else {
                hp->next = l2;
                l2 = l2->next;
            }
            hp = hp->next;
        }
        
        if (l1)
            hp->next = l1;
        if (l2)
            hp->next = l2;
        
        hp = rp;
        rp = rp->next;
        delete hp;
        return rp;
    }
};

File: algorithm/merge-two-sorted-lists/merge-two-sorted-lists.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
__date__ = ' 3 30, 2016 '
__author__ = 'mkfsn'

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} l1
    # @param {ListNode} l2
    # @return {ListNode}
    def mergeTwoLists(self, l1, l2):
        head = tail = ListNode(0)
        
        while l1 and l2:
            if l1.val < l2.val:
                node = ListNode(l1.val)
                head.next = node
                head = node
                l1 = l1.next
            else:
                node = ListNode(l2.val)
                head.next = node
                head = node
                l2 = l2.next
        
        if l1 is None:
            while l2:
                node = ListNode(l2.val)
                head.next = node
                head = node
                l2 = l2.next
        else:
            while l1:
                node = ListNode(l1.val)
                head.next = node
                head = node
                l1 = l1.next
        return tail.next
